api/src/controllers/changes.js

```diff
@@ -1,578 +1,396 @@
-    var result = {
-      id: change.id,
-      doc: change.doc
+  endFeed(feed);
+};
+
+const processChange = (change, seq) => {
+  let changeObj;
+  if (tombstoneUtils.isTombstoneId(change.id)) {
+    changeObj = {
+      change: tombstoneUtils.generateChangeFromTombstone(change),
+      viewResults: authorization.getViewResults(tombstoneUtils.extractDoc(change.doc))
     };
-    var row = getReplicationKey(change.doc);
-    if (row && row.length) {
-      result.subject = row[0];
-      result.submitter = row[1].submitter;
-    }
-    return result;
+  } else {
+    changeObj = {
+      change: change,
+      viewResults: authorization.getViewResults(change.doc)
+    };
+  }
+  // inform the normal feeds that a change was received while they were processing
+  normalFeeds.forEach(feed => {
+    feed.lastSeq = seq;
+    feed.pendingChanges.push(changeObj);
   });
-  continuousFeeds.forEach(function(feed) {
-    // check if new and relevant
-    if (hasNewApplicableDoc(feed, modifiedChanges)) {
-      abortUpstreamBatches(feed, 'newer-change-received', 'INFO [onRelevantChange] cancelling upstream batch (before bindServerIds()):');
-
-      bindServerIds(feed, function(err) {
-        abortUpstreamBatches(feed, 'newer-change-received', 'INFO [onRelevantChange] cancelling upstream batch (after bindServerIds()):');
-        if (err) {
-          return serverUtils.error(err, feed.req, feed.res);
-        }
-        feed.log('updateFeeds()', 'Calling getChanges()…');
-        getChanges(feed);
-      });
+
+  // send the change through to the longpoll feeds which are allowed to see it
+  longpollFeeds.forEach(feed => {
+    feed.lastSeq = seq;
+    if (!authorization.allowedChange(feed, changeObj)) {
+      feed.pendingChanges.push(changeObj);
+      return;
+    }
+
+    if (authorization.isAuthChange(feed, changeObj)) {
+      endFeed(feed, false);
+      processRequest(feed.req, feed.res, feed.userCtx);
+      return;
     }
+
+    addChangeToLongpollFeed(feed, changeObj);
   });
 };
 
-var init = function(since) {
+const initContinuousFeed = since => {
+  continuousFeed = db.medic
+    .changes({
+      live: true,
+      include_docs: true,
+      since: since || 'now',
+      timeout: false,
+    })
+    .on('change', (change, pending, seq) => {
+      processChange(change, seq);
+      currentSeq = seq;
+    })
+    .on('error', () => {
+      continuousFeed.cancel();
+      Promise.resolve().then(() => {
+        continuousFeed = initContinuousFeed(currentSeq);
+      });
+    });
+};
+
+const init = () => {
+  if (inited) {
+    return;
+  }
+  // As per https://issues.apache.org/jira/browse/COUCHDB-1288, there is a 100 doc
+  // limit on processing changes feeds with the _doc_ids filter within a
+  // reasonable amount of time.
+  // While hardcoded at first, CouchDB 2.1.1 has the option to configure this value
+  MAX_DOC_IDS = config.get('changes_doc_ids_optimization_threshold') || MAX_DOC_IDS;
+  db.medic.setMaxListeners(100);
```

Comment:
Do we need this? Does this increase a limit or set a limit? How many listeners do we have?

Reply:
PouchDB DB instances are EventEmitters. By default, node allows 10 listeners/EventEmitter. When the 11th is added, it issues a warning (https://nodejs.org/api/events.html#events_emitter_setmaxlisteners_n). 
The `changes` call is also an EventEmitter, but it also adds a listener to the main db `destroyed` event. Because we issue the concurrent requests with `_doc_ids` filter, we can easily pollute the log with many of these warnings. 

I guess I could move this line of code in the /db-pouch.js file where the db.medic object is constructed and relieve the limit altogether (allow infinite listeners).

Reply:
Yeah maybe move it to db-pouch. Were we getting lots of errors with the old implementation? I thought this implementation would have fewer listeners than the old one.

Reply:
Ahh of course, that makes sense.